File: csharp/main/LiteralUInt32.cs
using System;

namespace Deltix.Luminary
{
	public sealed class LiteralUInt32 : LiteralInteger, ILiteral<UInt32>
	{
		public LiteralUInt32(UInt32 value)
			: base(LiteralKind.UInt32)
		{
			Value = value;
		}

		public UInt32 Value { get; }

		public override LiteralInteger CastTo(IntegralType type)
		{
			switch (type)
			{
				case IntegralType.Int64:
					return new LiteralInt64(Value);

				case IntegralType.UInt64:
					return new LiteralUInt64(Value);

				case IntegralType.Int32:
					if (Value > Int32.MaxValue)
						throw new InvalidCastException($"Value '{Value}' cannot be safely cast to 'Int32'.");
					return new LiteralInt32((Int32) Value);

				case IntegralType.UInt32:
					return this;

				case IntegralType.Int16:
					if (Value > Int16.MaxValue)
						throw new InvalidCastException($"Value '{Value}' cannot be safely cast to 'Int16'.");
					return new LiteralInt16((Int16) Value);

				case IntegralType.UInt16:
					if (Value > UInt16.MaxValue)
						throw new InvalidCastException($"Value '{Value}' cannot be safely cast to 'UInt16'.");
					return new LiteralUInt16((UInt16) Value);

				case IntegralType.Int8:
					if (Value > SByte.MaxValue)
						throw new InvalidCastException($"Value '{Value}' cannot be safely cast to 'Int8'.");
					return new LiteralInt8((SByte) Value);

				case IntegralType.UInt8:
					if (Value > Byte.MaxValue)
						throw new InvalidCastException($"Value '{Value}' cannot be safely cast to 'UInt8'.");
					return new LiteralUInt8((Byte) Value);

				default:
					throw new ArgumentOutOfRangeException(nameof(type), type, null);
			}
		}

		public override String ToString()
		{
			return $"{Value}u32";
		}
	}
}

File: java/src/main/java/deltix/luminary/LiteralUInt32.java
package deltix.luminary;

public class LiteralUInt32 extends LiteralInteger {
    private int value;

    public LiteralUInt32(int value) {
        super(LiteralKind.UINT32);
        this.value = value;
    }

    public int getValue() {
        return value;
    }

    @Override
    public LiteralInteger castTo(IntegralType type) {
        long longValue = value & 0xFFFFFFFFL;
        switch (type) {
            case INT8:
                if (longValue > Byte.MAX_VALUE)
                    throw new ClassCastException("Value '" + longValue + "' cannot be safely cast to 'Int8'.");
                return new LiteralInt8((byte) longValue);

            case UINT8:
                if (longValue > 0xFF)
                    throw new ClassCastException("Value '" + longValue + "' cannot be safely cast to 'UInt8'.");
                return new LiteralUInt8((byte) longValue);

            case INT16:
                if (longValue > Short.MAX_VALUE)
                    throw new ClassCastException("Value '" + longValue + "' cannot be safely cast to 'Int16'.");
                return new LiteralInt16((short) longValue);

            case UINT16:
                if (longValue > 0xFFFF)
                    throw new ClassCastException("Value '" + longValue + "' cannot be safely cast to 'UInt16'.");
                return this;

            case INT32:
                if (longValue > Integer.MAX_VALUE)
                    throw new ClassCastException("Value '" + longValue + "' cannot be safely cast to 'Int32'.");
                return new LiteralInt32((int) longValue);

            case UINT32:
                return this;

            case INT64:
                return new LiteralInt64(longValue);

            case UINT64:
                return new LiteralUInt64(longValue);

            default:
                throw new IllegalArgumentException("Undefined integral type.");
        }
    }

    @Override
    public String toString() {
        return Long.toString(value & 0xFFFFFFFFL) + "u32";
    }
}
